src/broadcast/broadcast.rs

```diff
@@ -252,7 +253,7 @@ impl<N: NodeIdT> Broadcast<N> {
     /// Handles a received `Echo` message.
     fn handle_echo(&mut self, sender_id: &N, p: Proof<Vec<u8>>) -> Result<Step<N>> {
         // If the sender has already sent `Echo`, ignore.
-        if let Some(old_p) = self.echos.get(sender_id) {
+        if let Some((_, Some(old_p))) = self.echos.get(sender_id) {
```

Comment:
Actually, even if this is `Some((hash, None))`, we should probably check that `hash` agrees with `p.root_hash()`: If it doesn't, the sender must be faulty, too.

Reply:
You mean same sender sent different `hash` with `Echo` and `EchoHash` messages. Makes sense.